### PR TITLE
Hadriens/seg metrics per class

### PR DESCRIPTION
**Pull request by Farbum**
This PR adds:

- Dice loss for segmentation task
- X-entropy optional weights for class imbalance
- optional reporting of metric per class for segmentation metrics

### CODE REVIEW

File: rslearn/train/tasks/segmentation.py [hunk 1]
```diff
@@ -332,7 +406,7 @@ def __init__(
         self,
         metric: Metric,
         pass_probabilities: bool = True,
-        class_idx: int | None = None,
```
Comment (favyen2) [inline rslearn/train/tasks/segmentation.py]: It seems like this will break existing configs such as https://github.com/allenai/olmoearth_projects/blob/b9329999162a6265566cf0f3e1fbc66454194577/olmoearth_run_data/nandi/model.yaml#L67

Is it possible to deprecate the class_idx setting instead of removing it entirely? Or keep it because it could be useful if users want to apply various torchmetrics which do return a tensor instead of dict (so should be indexed by integer, not string).

File: rslearn/train/tasks/segmentation.py [hunk 2]
```diff
@@ -584,9 +658,11 @@ def compute(self) -> Any:
         """Compute metric.
 
         Returns:
-            the mean IoU across classes.
+            dict with "avg" containing the mean IoU across classes, and
```
Comment (favyen2) [inline rslearn/train/tasks/segmentation.py]: Do you think it'd be worthwhile to follow the convention in torchmetrics like https://lightning.ai/docs/torchmetrics/stable/classification/precision.html#torchmetrics.classification.MulticlassPrecision and have argument to the constructor to control the averaging? Then it can either return a scalar mean IoU here or a 1D tensor of per-class mean IoUs. And the SegmentationMetric can work with both built-in metrics and the custom ones we've defined here.

File: rslearn/train/data_module.py [hunk 3]
```diff
@@ -231,6 +231,7 @@ def _get_dataloader(
                 num_replicas=self.trainer.world_size,
                 rank=self.trainer.global_rank,
                 shuffle=should_shuffle,
+                drop_last=(split == "val"),  # Ensure even batches for validation
```
[the remark this answers is not in the document]
Comment (Farbum) [reply rslearn/train/data_module.py]: I was about to remove this line. It fixes a bug that occurs with multi-gpu fine-tuning.
This is part of a bigger problem that I am exploring in a different PR, so I am taking it off this PR.

File: rslearn/train/tasks/segmentation.py (hunk 1 above)
Comment (Farbum) [reply rslearn/train/tasks/segmentation.py]: Good point, added the logic to keep both

File: rslearn/train/tasks/segmentation.py (hunk 2 above)
Comment (Farbum) [reply rslearn/train/tasks/segmentation.py]: Ok I read your comment too fast.
So unfortunately lightning does not handle nested dictionaries (a metric collection item returning a dictionary). 
This is why I opted first to create one metric object per metric per class.

That being said, there is a way to do it that requires minor changes to lightning and multi_task.
- lightning has to first compute the metrics via the collection (which flattens the nest dict), then we log. That requires to move the logging from validation_step to on_validation_epoch_end and manually compute/reset the metrics.
- multi_task needs to flatten the structure so there's only one MetricCollection with all metrics at the top level.

I tested it and it works. Not too sure which option is cleanest, though.
What do you think?
